### ISSUE TITLE
Salesforce CLI Scanner Integration

### ISSUE THREAD

**Issue by jonathanwiesel** [forcedotcom/salesforcedx-vscode#2724]
**Is your feature request related to a problem? Please describe.**
It's common in a development workflow to use static analysis tools that search the code for problems before proceeding to make it available for usage, this way you can guarantee the code follows a certain degree of security measures and best practices by using standard rules in a project / company.

**Describe the solution you'd like**
With the recent introduction of the [Salesforce CLI Scanner Plugin](https://forcedotcom.github.io/sfdx-scanner/) seems like the only thing missing is filling the gap between the development environment (VS Code) and the plugin. 

It should have the ability to run the static analysis on demand directly from the VS Code palette, or automatically when a file is saved. Results should be shown as errors, warnings or hints in the Problems tab.

**Describe alternatives you've considered**
There's currently a [great extension (Apex PMD)](https://marketplace.visualstudio.com/items?itemName=chuckjonas.apex-pmd) developed by @ChuckJonas that fills the gap between PMD and Apex and shows the output appropriately, however seems like the logical way moving forward is to delegate all the communication with the static analysis tools to the CLI Scanner plugin and make the extension interact directly with the plugin instead making the implementation on the extension decoupled from the engine, rules and configuration.

**Comment by ntotten** [comment]
Given that the CLI scanner just uses PMD and ESLint under the hood I am not really sure what value this have beyond just using the PMD and ESLint extensions. Could you explain more on why we would need to make a special extension for this?

**Comment by jonathanwiesel** [comment]
Good point @ntotten, since the plugin offers a language/engine independent approach of rule management, execution and output you have a unified interface of communication between VS Code and the static analysis tools, you don't need to manage different configuration types for each different language and therefore each extension (in such case it will be managed in the plugin itself). 

Basically you only need one extension (and obviously the one plugin behind it) to manage al the heavy-lifting. 
Also, as the plugin evolves and introduces more engines and more capabilities, a new extension and a new way of interacting with it won't be necessary, just adjustments of the existing one to, for example, match the new language engine to the file definitions.

**Comment by ChuckJonas** [comment]
IMO, the approach putting all your static analysis into one bag is great for CI processes, but not ideal when you're talking about a vscode extension.

Sure, a command to scan an entire project and create vscode `Diagnostics` wouldn't hurt, but also not all that useful when you are actively developing.  The Scanner seems to be optimized to run at the project level vs something like `onDocumentChange`.   It's really built to be incorporated into a git hook or in your CI.

In general, specialized extensions are able to offer more and are easier to extend than a generic, all encompassing one.  

IOW, I don't need to see eslint errors from the sf scanner, when I can "fix them on save" using the official extension...

I agree these tools should share configuration, but that should be achieved by the scanner taking into consider the individual config files in the workspace (`eslintrc.json`, `ruleset.xml`, etc).  

Unfortunately, it looks like they are depreciating the ability to pass in a `ruleset`, so that will make it tougher to keep it in sync with Apex PMD (unless I implement the "Category" approach, which honestly isn't granular enough IMO).

Just my 2 cents

**Comment by msrivastav13** [comment]
@ChuckJonas We have conveyed this to the team that built the scanner. They will address the ability to use ruleset.xml and all other things requested here.

**Comment by lcampos** [comment]
This could be folded into this other issue on the scanner repo https://github.com/forcedotcom/sfdx-scanner/issues/252. Will leave it up to them to decide.
